engine/table/src/main/java/io/deephaven/engine/table/impl/util/ShiftInversionHelper.java

```diff
@@ -10,31 +10,47 @@
  */
 public class ShiftInversionHelper {
 
-    final RowSetShiftData shifted;
+    final private RowSetShiftData shifted;
+    final private boolean reverseOrder;
 
-    private int destShiftIdx = 0;
+    private int destShiftIdx;
 
     public ShiftInversionHelper(final RowSetShiftData shifted) {
+        // if not specified, assume forward viewport ordering
+        this(shifted, false);
+    }
+
+    public ShiftInversionHelper(final RowSetShiftData shifted, final boolean reverseOrder) {
         this.shifted = shifted;
+        this.reverseOrder = reverseOrder;
+        this.destShiftIdx = reverseOrder ? shifted.size() : 0;
     }
 
     private void advanceDestShiftIdx(final long destKey) {
         Assert.geq(destKey, "destKey", 0);
-        destShiftIdx = (int) binarySearch(destShiftIdx, shifted.size(), innerShiftIdx -> {
-            long destEnd = shifted.getEndRange((int) innerShiftIdx) + shifted.getShiftDelta((int) innerShiftIdx);
-            // due to destKey's expected range, we know this subtraction will not overflow
-            return destEnd - destKey;
-        });
+        destShiftIdx = (int) binarySearch(
+                reverseOrder ? 0 : destShiftIdx,
```

Comment:
Should this be `-1` instead of `0`?

Reply:
Using `reverseOrder ? -1 : destShiftIdx` results in an infinite loop in `binarySearch()`